includes/class-sensei-quiz.php

```diff
@@ -1829,6 +1942,23 @@ public function update_quiz_author( int $quiz_id, int $new_author_id ) {
 			Sensei()->question->maybe_update_question_author( $question->ID, $new_author_id );
 		}
 	}
+
+	/**
+	 * Replace all pagination links with buttons (<a> => <button>).
+	 *
+	 * @since 3.15.0
+	 *
+	 * @param string $html The pagination html.
+	 *
+	 * @return string
+	 */
+	public function replace_pagination_links_with_buttons( $html ): string {
```

Comment:
I took a closer look at this PR to better understand the problem. Converting the links to buttons does feel somewhat hacky. I'm curious as to why a JavaScript solution was dismissed? It seems like it might be a bit cleaner.

Reply:
I think the main reason we went with this method is for simplicity - it uses the same state save/submit mechanism as the non-paginated form and avoids all the potential pitfalls.

If we go the JS way, we need to consider the following:

 * Are we OK with the form working differently compared to the non-paginated form? Are we OK with maintaining it?
 * Should we save the state using AJAX calls each time there is a form change or put the answers in localStorage (sessionStorage?) and send that on submit?
 * Are we doing a page refresh between pages?
 * If we use AJAX calls, we should consider using debouncing and avoiding request race conditions.
 * How will the file submission work? Maybe that will be a problem if using localStorage.
 * How to preserve the questions order between pages when the questions are randomized? One solution here would be to save the questions order initially, but doing so will make the form work differently compared to the non-paginated form.
 * Could we paginate the form entirely using JS? Not a big fan of this, but should be very simple to do. We query all the questions each time anyway.
 * How will this solution hold in the future? For example, what will happen if we add backend validation or more complex question types?

@gkaragia What's your opinion?

Reply:
In my opinion, we could use the AJAX on-change method to save the state. Most of the code from this PR can be used for the implementation and it's somewhat simple to do if we figure out how to deal with the randomized questions order. Thoughts?

Reply:
It sounds like you've already thought this through very well. I was mostly wanting to understand the difficulties with a JS-based approach, and you've explained the challenges with it, so thank you! Given that you are so much closer to the problem, it sounds good to me to continue on with your current approach.

Reply:
@donnapep I was thinking about the links-to-buttons hack and came up with a potentially better solution:

* Keep the pagination as links.
* Hook to the `beforeunload` JS event and show a "Changes that you made may not be saved" popup message when the user has made changes to the form and tries to navigate away or go to another page using the links and not the submit buttons. This approach is the same as what WordPress is going in the admin and I think it's a good user experience.
* When having randomized questions - save the questions order on the first load. That way the questions are randomized once and not on every page. I guess this will change the current behavior, but it makes sense to me. It's strange to randomize the questions each time the quiz is refreshed because this can be abused by the quiz taker to land on questions that he actually knows.

Thoughts? :slightly_smiling_face:

Reply:
> Hook to the `beforeunload` JS event and show a "Changes that you made may not be saved" popup message when the user has made changes to the form and tries to navigate away or go to another page using the links.

Hmmm...interesting idea. I suppose the concern here would be that showing a popup isn't a technique we currently use for displaying messages elsewhere in the frontend. We typically work with Sensei's existing system for displaying notices.

Assuming we did show a dialog, what action would the user need to take in order to save? Something like cancel the dialog, click on the _Save_ link, and then navigate to another page? Would they get the message again at that point? If so, I could see something like this being a bit disruptive to the overall flow. My preference would be to sacrifice a bit on the technical side in order to create a better experience for the user.

> When having randomized questions - save the questions order on the first load.

This sounds like a good idea. How would we handle questions that may have been removed or added to the quiz since the user first viewed it? We'd need to account for those somehow, maybe not so much for questions that have been added, but certainly for ones that were removed.

Reply:
> Assuming we did show a dialog, what action would the user need to take in order to save?

I've made a [quick sample](https://jsfiddle.net/yLscnj8m/) of how this would work. Basically, the user would have to hit cancel and submit.

I guess a better solution would be to save the form on each change (autosave feature?). Similar to GitHub, Trello, etc.

>My preference would be to sacrifice a bit on the technical side in order to create a better experience for the user.

I'm OK with this too. 👍 

> This sounds like a good idea. How would we handle questions that may have been removed or added to the quiz since the user first viewed it?

It would work the same way as if you hit the save button. Currently, if a user saves a quiz and a question block is removed, that question will still be displayed to that user. They need to hit the reset in order to refresh the questions. The same thing with new questions, they won't be displayed until reset.

Reply:
> I've made a quick sample of how this would work. Basically, the user would have to hit cancel and submit.

Hmm...yeah, I think that's too much friction. Auto-saving seems a better UX.

> Currently, if a user saves a quiz and a question block is removed, that question will still be displayed to that user.

OK, if that's how it works currently then that sounds a-ok. 👍🏻